New PSBT global keys

dr-orlovsky opened the pull request:
Based/depends on https://github.com/rust-bitcoin/rust-bitcoin/pull/470

Part of epic https://github.com/rust-bitcoin/rust-bitcoin/issues/473

stevenroose commented on src/util/psbt/map/global.rs:
```diff
@@ -154,6 +164,58 @@ impl Decodable for Global {
+                                let mut decoder = Cursor::new(pair.value);
+                                let mut fingerprint = [0u8; 4];
+                                decoder.read_exact(&mut fingerprint[..])?;
+                                let mut path: Vec<ChildNumber> = Default::default();
```
Hmm, about this whole section, it seems a bit strangely setup.

Basically after reading the fingerprint, the number of ChildNumbers is `decoder.len() / 4`, right? You can error early if `decoder.len() % 4 != 0` or even `pair.value.len() % 4 != 0` because the fingerprint is also 4 bytes.

And then `path` can be `Vec::with_capacity(decoder.len() / 4)` and then you can use while let for the parsing:
```rust
while let Ok(index) = u32::consensus_decode(&mut decoder) {
    path.push(index.into());
}
```

dr-orlovsky replied:
Addressed with the new commit

stevenroose reviewed (changes requested):
You also forgot to implement `Map::get_pairs` to do the serialization part.

Some unit tests that add some roundtrips of some PSBTs with the new keys would also be appreciated.

apoelstra commented:
needs rebase

dr-orlovsky commented:
Rebased and addressed @stevenroose comments. NB: this is based on #470, which also had review comments (they are addressed now, but #470 needs to be merged first)

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -572,7 +644,49 @@ impl ExtendedPubKey {
+        let mut ret = [0; 78];
+        ret[0..4].copy_from_slice(&match self.network {
+            Network::Bitcoin => [0x04u8, 0x88, 0xB2, 0x1E],
+            Network::Testnet | Network::Signet | Network::Regtest => [0x04u8, 0x35, 0x87, 0xCF],
```
The BIP only specifies versions for Testnet, is it safe to use the same for Signet and Regtest

apoelstra replied:
I think it's reasonable for now. Regtest copies a lot of values from testnet.

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -531,16 +531,14 @@ impl ExtendedPrivKey {
-                ).map_err(|e|
-                    base58::Error::Other(e.to_string())
-                )?,
+                ).expect("correct size"),
```
Can the `from_slice()` only fail if size is incorrect? 
It looks like https://github.com/bitcoin-core/secp256k1/blob/9e5939d2848df5508e3a6eb038273b28d255248c/src/secp256k1.c#L550-L559 can fail if there is overflow or if secret key is zero.

dr-orlovsky replied:
Fixed in the other branch and now fine after the rebase

dr-orlovsky commented on src/util/bip32.rs:
```diff
@@ -388,6 +388,12 @@ pub enum Error {
     InvalidDerivationPathFormat,
+    /// Unknown version magic bytes
+    UnknownVersion([u8; 4]),
+    /// Encoded extended key data have wrong length
```
Fixed this one and below with https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/9f2ca1428cef2091cae043aa483439878f236936

apoelstra commented:
Can you squash these commits together? It's not too long to read, and it's a bit harder to review because you make changes that are later modified.

justinmoon commented on src/util/psbt/map/global.rs:
```diff
@@ -155,6 +165,55 @@ impl Decodable for Global {
                                 return Err(Error::InvalidKey(pair.key).into())
                             }
                         }
+                        0x01 => {
```
Might be nice to have a comment about which PSBT field these hex values represent

justinmoon replied:
Also, I think you should change `0u8` to `0x00` on line 140 above

justinmoon commented:
I think this breaks PSBT serialization.

The `impl_psbtmap_consensus_encoding` macro calls `Global::get_pairs` which needs to be updated after you removed `version` and `xpub` values from `Global::unknown` and put them in their own fields.

[Here's a failing test](https://gist.github.com/justinmoon/b8274a4162c9b8713559d7f37af3073a).

dr-orlovsky commented:
@justinmoon you are right, and even more: we need to do a proper merge according to BIP174 rules and resolve conflicts. Working on the change

dr-orlovsky commented:
@apoelstra squashed old commits; new commits are on top and they do not change previous code but splits added functionality into logical blocks for review simplification

In general, it was not that trivial to do the merge procedure and key pair export; there were two questions which I do not know the right answer for, but proposing what I think will work the best:
1. What to do when two PSBTs global xpub match by value, but not derivation/key source? Spec says that it's up to an implementation to decide, so we have to decide :)
2. Do we need to serialize default 0 version, or just ignore this case? If we will serialize, this will break test vectors; if not, it may break roundtrips if the original PSBT contained explicit 0-version global key

My current implementation:
1. Uses special multi-step algorithm described here: https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/8b672c348e8cac8dd936353a641f8db0afdf0632#diff-2581ef26b841ed4069017f0e46e1393216bf6b413302ae34f248105526c8755bR137-R164
2. Does not serialize default 0 version as a separate key.

apoelstra commented:
1. I'm happy with what you've done. I would've gone even simpler and just taken the lexicograpically first one :) but you're right that probably "longest first" is likely to do the right thing in realistic conflict scenarios.
2. Agreed, we'll just break round-tripping (and will have to add extra canonicalization to our fuzztests)

justinmoon commented on src/util/psbt/map/global.rs:
```diff
@@ -92,6 +106,34 @@ impl Map for Global {
+                },
+                value: {
+                    let mut ret = Vec::with_capacity(4 + derivation.len() * 4);
+                    ret[0..4].copy_from_slice(fingerprint.as_bytes());
```
This will cause runtime error. `ret` is a vec with length 0, so you can't use `vec[0..4]`

dr-orlovsky replied:
Fixed

stevenroose commented:
You seem to have some Rust v1.29 compilation error in the ordering algorithm.

dr-orlovsky commented:
@stevenroose @apoelstra 1.29 fixed for all commit history. It's extremely stupid with borrow checks and making it compiling required an unnecessary clone; I have tried everything else and it simply was not working with 1.29 :(

apoelstra commented on src/util/psbt/map/global.rs:
```diff
@@ -118,6 +121,54 @@ impl Map for Global {
+                    // 4) choose largest fingerprint; if equal
+                    // 5) do nothing
+                    // Clone is required for Rust 1.29 borrow checker
+                    let (fingerprint2, derivation2) = entry.get().clone();
```
You can replace this with
```
                    let (fingerprint2, len2, normal_len2) = {  // weird scope needed for rustc 1.29 borrowck
                        let (fp, deriv) = entry.get().clone();
                        (fp, deriv.len(), deriv.into_iter().rposition(ChildNumber::is_normal))
                    };
```

dr-orlovsky replied:
Done

apoelstra commented on src/util/psbt/map/global.rs:
```diff
@@ -110,6 +111,8 @@ impl Map for Global {
         Ok(rv)
     }
 
+    // Keep in mind that according to BIP 174 this function must be transitive, i.e.
```
nit: the correct term is "commutative"

dr-orlovsky replied:
Fixed

apoelstra reviewed (commented):
ack 794c5196a0eb4ec635a35cf09708ff8da6a00d28

apoelstra reviewed (dismissed):
ack 794c5196a0eb4ec635a35cf09708ff8da6a00d28

stevenroose commented on src/util/psbt/map/global.rs:
```diff
@@ -110,6 +154,55 @@ impl Map for Global {
+                    // 1) if fingerprints are equal but derivations are not, we report an error; otherwise
+                    // 2) choose longest unhardened derivation; if equal
+                    // 3) choose longest derivation; if equal
+                    // 4) choose largest fingerprint; if equal
```
Largest here means just by ordering the bytes, right? So 0xdeadbeef wins from 0xbeadbeef? This is a bit arbitrary. But I suppose I can agree that it's better to do something then to error.

dr-orlovsky replied:
Correct. Actually since fingerprint is `[u8; 4]` and not `u32` I assume it uses little-endian comparison. And yes, this is arbitrary

stevenroose replied:
nit: I'd perhaps rephrase "choose largest pingerprint using binary ordering" because this suggests that the fingerprint "size" might have real meaning :p

stevenroose reviewed (approved):
utACK d1ace733e0de532707df7140ee5be1eb3faedbe0

stevenroose commented:
@apoelstra you "approved" a different commit than the hash you mention inside the approve message. Can you confirm you meant to ack https://github.com/rust-bitcoin/rust-bitcoin/commit/d1ace733e0de532707df7140ee5be1eb3faedbe0?

stevenroose commented on src/util/bip32.rs:
```diff
@@ -126,14 +126,14 @@ impl ChildNumber {
     ///
     /// [`Normal`]: #variant.Normal
-    pub fn is_normal(self) -> bool {
+    pub fn is_normal(&self) -> bool {
```
nit: There's no need to do this. It's probably microscopically more efficient to leave it `self`.

dr-orlovsky replied:
This is not about efficiency, this is required by compiler, otherwise we get

```console
error[E0631]: type mismatch in function arguments
   --> src/util/psbt/map/global.rs:182:71
    |
182 |                         (fp, deriv.len(), deriv.into_iter().rposition(ChildNumber::is_normal))
    |                                                                       ^^^^^^^^^^^^^^^^^^^^^^ expected signature of `fn(&ChildNumber) -> _`
    | 
   ::: src/util/bip32.rs:129:5
    |
129 |     pub fn is_normal(self) -> bool {
    |     ------------------------------ found signature of `fn(ChildNumber) -> _`

```

apoelstra replied:
you can just add `.cloned()` on the iterator before `.rposition`. Or `.copied()` in Rust 1.36. But what you did here is fine.

stevenroose commented on src/util/bip32.rs:
```diff
@@ -126,14 +126,14 @@ impl ChildNumber {
     ///
     /// [`Hardened`]: #variant.Hardened
-    pub fn is_hardened(self) -> bool {
+    pub fn is_hardened(&self) -> bool {
```
nit: There's no need to do this. It's probably microscopically more efficient to leave it `self`.

dr-orlovsky replied:
Same as above

stevenroose commented on src/util/psbt/map/global.rs:
```diff
@@ -92,6 +106,34 @@ impl Map for Global {
+                value: {
+                    let mut ret = Vec::with_capacity(4 + derivation.len() * 4);
+                    ret.extend(fingerprint.as_bytes());
+                    for no in 0..derivation.len() {
```
this can be `for n in &derivation` without the index. Or even 
```rust
derivation.for_each(|n| ret.extend(&u32_to_array_le(n.into())));
```

dr-orlovsky replied:
No, it can't be: there is no `iter()` for `Derivation`. You can only use `derivation.into_iter()`

stevenroose commented on src/util/psbt/map/global.rs:
```diff
@@ -110,6 +154,55 @@ impl Map for Global {
+        //          the specification. It can pick arbitrarily when conflicts occur.
+
+        // Keeping the highest version
+        self.version = self.version.max(other.version);
```
nit: I think `cmp::max(self.version, other.version)` reads better; but very much a nit :)

stevenroose commented on src/util/psbt/map/global.rs:
```diff
@@ -110,6 +154,55 @@ impl Map for Global {
+                        (Ordering::Equal, _, Ordering::Equal) => {
+                            return Err(psbt::Error::MergeConflict(format!(
+                                "global xpub {} has inconsistent key sources",
+                                xpub
```
nit: Put on same line as string as long as it doesn't exceed 100 width :)

stevenroose commented on src/util/psbt/map/global.rs:
```diff
@@ -155,6 +250,55 @@ impl Decodable for Global {
+                            if !pair.key.key.is_empty() {
+                                let xpub = ExtendedPubKey::decode(&pair.key.key)
+                                    .map_err(|_| {
+                                        encode::Error::ParseFailed("Can't deserialize ExtendedPublicKey from global XPUB key data")
```
nit: This exceeds 100 characters, right? how about putting the encode::Error::ParseFailed on the same line as `.map_err`?

stevenroose commented on src/util/psbt/map/global.rs:
```diff
@@ -155,6 +250,55 @@ impl Decodable for Global {
+                                let keys_count = pair.value.len() / 4 - 1;
+                                let mut decoder = Cursor::new(pair.value);
+                                let mut fingerprint = [0u8; 4];
+                                decoder.read_exact(&mut fingerprint[..])?;
```
nit: I think you can actually do
```rust
let mut fingerprint = Fingerprint::default();
decoder.read_exact(&mut fingerprint[..])?;
```
but I'm not 100% sure

dr-orlovsky replied:
You can't: there is no `IndexMut` impl for `Fingerprint`

stevenroose reviewed (commented):
Left a bunch of nits. I didn't want to make them to not hold up the MR, but since your CI fails and you need to push changes anyway, I thought I'd leave them anyway.

dr-orlovsky commented:
Had to rebase since it had conflicts with merged #530 and had a failed fuzz tests. Also used this opportunity to address @stevenroose nits. Will need a re-ACK from you guys, @apoelstra, @stevenroose

sgeisler commented on src/util/psbt/map/global.rs:
```diff
@@ -158,6 +250,55 @@ impl Decodable for Global {
+                                        return Err(encode::Error::ParseFailed("Wrong global version value length (must be 4 bytes)"))
+                                    }
+                                    version = Some(Decodable::consensus_decode(&mut decoder)?);
+                                    if decoder.position() != vlen as u64 {
```
Why would a decoder for an `u32` _not_ consume all 4 bytes? The check seems redundant.

dr-orlovsky replied:
Fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/b84faa7f5e6f9d273a4a6cb78098f10c73594783

sgeisler commented on src/util/psbt/map/global.rs:
```diff
@@ -158,6 +250,55 @@ impl Decodable for Global {
+                                    if vlen != 4 {
+                                        return Err(encode::Error::ParseFailed("Wrong global version value length (must be 4 bytes)"))
+                                    }
+                                    version = Some(Decodable::consensus_decode(&mut decoder)?);
```
We only understand version 0 PSBTs. According to BIP-174 we should throw an error if we see anything other than version 0.

> "If a parser encounters a version number it does not recognize, it should exit immediately as this indicates that the PSBT will contain types that it does not know about and cannot be ignored." -- ([Version Numbers](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#version-numbers))

dr-orlovsky replied:
Fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/b84faa7f5e6f9d273a4a6cb78098f10c73594783

apoelstra commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,54 @@ impl Map for Global {
+                    let len1 = derivation1.len();
+                    let normal_len1 = derivation1.into_iter().rposition(ChildNumber::is_normal);
+                    match (normal_len1.cmp(&normal_len2), len1.cmp(&len2), fingerprint1.cmp(&fingerprint2)) {
+                        (Ordering::Equal, Ordering::Equal, Ordering::Equal) => {},
```
Their lengths are equal and their fingerprints are equal. How else could they differ?

apoelstra replied:
Ah, as you observe below it's pretty easy to collide fingerprints. That's fair.

sgeisler replied:
Not only that. But you also just compare derivation path length. Assuming the same fingerprint/key the both derivation paths `m/0'/0` and `m/1'/2` denote different keys but have the same `normal_len` and `len`, which is all that's being tested for if I'm not mistaken.

dr-orlovsky replied:
Fingerprints can't collide here, since if we also have xpub equality. Basically I will add comparison between full derivation paths, and if they are equal it will mean that we have xpubs equal, derivation paths equal and equal fingerprints will mean equality, it can't be a collision anymore.

sgeisler replied:
Oh, right, I didn't take into account that the xpub has to be equal in the first place.

sgeisler replied:
Or wait, that assumes that the fingerprint+derivation path has to actually lead to the same pubkey, but what about attacker-provided PSBTs where that isn't the case? Can anything bad happen in that case?

dr-orlovsky replied:
Fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/21c11e3315d6753b868abc7a61a0772c46f5646c

dr-orlovsky commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,54 @@ impl Map for Global {
+                    let normal_len1 = derivation1.into_iter().rposition(ChildNumber::is_normal);
+                    match (normal_len1.cmp(&normal_len2), len1.cmp(&len2), fingerprint1.cmp(&fingerprint2)) {
+                        (Ordering::Equal, Ordering::Equal, Ordering::Equal) => {},
+                        (Ordering::Equal, _, Ordering::Equal) => {
```
Fixing with derivation path comparison

dr-orlovsky replied:
Fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/21c11e3315d6753b868abc7a61a0772c46f5646c

dr-orlovsky commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,54 @@ impl Map for Global {
+                },
+                Entry::Occupied(mut entry) => {
+                    // Here in case of the conflict we select the version with algorithm:
+                    // 1) if fingerprints are equal but derivations are not, we report an error; otherwise
```
avoiding that with making sure derivation paths are per-item equal

dr-orlovsky replied:
Fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/21c11e3315d6753b868abc7a61a0772c46f5646c

dr-orlovsky commented on src/util/psbt/map/global.rs:
```diff
@@ -158,6 +250,55 @@ impl Decodable for Global {
+                                    return Err(encode::Error::ParseFailed("Incorrect length of global xpub list"))
+                                }
+
+                                let keys_count = pair.value.len() / 4 - 1;
```
Fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/b84faa7f5e6f9d273a4a6cb78098f10c73594783

sgeisler commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,60 @@ impl Map for Global {
+                        }
+                        (Ordering::Greater, ..)
+                        | (Ordering::Equal, Ordering::Greater, ..)
+                        | (Ordering::Equal, Ordering::Equal, Ordering::Greater, _) => {
```
I still have the feeling there's  an "inconsistent key source" hiding in here. Does it ever make sense that the same key can be derived from two different parents using equal-length derivation paths? That would make the key-ancestry a DAG instead of a tree, which seems odd (haven't checked the crypto, maybe it could be constructed, but I doubt it).

I guess specifying what we actually want to achieve instead of how we get there would be helpful. From what I gather it would be along the lines of "always choose the most senior key in the ancestry-chain, fail otherwise". 

That would mean checking:
* if fingerprint and derivation path are equal do nothing
* else if one (naturally the shorter) derivation path is a suffix of the other choose the entry of the longer derivation path
* else return an error, something is fishy

Sorry for coming up with this so late, I hope it makes sense. If not I'm not too opposed to potentially not detecting all key source inconsistencies, seems like any serious interactive protocol needs its own much stricter validation before merging potentially attacker-generated PSBTs. But we should be clear on that.

apoelstra replied:
No, if you had any "merges" in the key tree that would imply a hash collision.

The only way to get differing paths is for one path to be a truncation of the other (with a prefix collapsed into the fingerprint), which would be visible in the length.

dr-orlovsky replied:
Sorry, have not understood do I need to change something here or not. If not, @sgeisler can you pls re-review the PR such that it does not require my changes?

sgeisler replied:
This needs changes: `(Ordering::Equal, Ordering::Equal, Ordering::Greater, _)` means that two different derivation paths of the same length exist for the same key, which can't be, so this should probably be an error case.

My broader observation is that this logic is unnecessarily convoluted, making it prone to such mistakes, which is why I propose the simpler algorithm outlined above. I think a simpler algorithm with essentially the same, but now clearly stated goal, would be preferable here but I won't block the PR because of that.

dr-orlovsky replied:
Rewrote algorithm taking you idea as a baseline and simplifying it a bit further.

Also, I think that we have to choose longest **hardened** path, since it contain more of valuable information

sgeisler replied:
I don't fully understand what you mean with "contain more of valuable information".

If two paths lead to the same key one needs to be a suffix of the other. The only question is which of the both to choose: the longer or the shorter one. Under which circumstance would you choose the shorter one and under which the longer? I guess you mean to make the choice somehow based on hardened derivations in the prefix (part not contained in short path) of the longer path? If so what's the motivation?

dr-orlovsky replied:
If we have two paths like `[ac04]/48'/1'/1'/3` and `[8b9a]/1'/3` the first version presents more information since it is not just longer, but also contains more hardened keys in it.

Writing the example I think I understood what you mean: that if for two derivation paths of different lengths the shortest one is not a strict prefix of the other, something is wrong. I see... Let me modify the logic, it will simplify it even more.

sgeisler replied:
Yeah, I think given this fact it's probably a sane (not always the best, but you can't please everyone and need to standardize _something_) choice to always take the longer one.

dr-orlovsky replied:
@sgeisler pushed reworked version

sgeisler reviewed (commented):
I think by and large this PR is in a good state. I'm still not super happy with the merge logic, it seems way too convoluted, reasoning about it was somewhat hard and I'm not sure it's entirely consistent. But since many reviewed so far and didn't take issue with it I'd be fine with it since the BIP didn't specify to detect such inconsistencies afaik. @apoelstra, what do you think, keep polishing or merge?

dr-orlovsky commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,60 @@ impl Map for Global {
+                            fp,
+                            deriv.len(),
+                            deriv.into_iter().rposition(ChildNumber::is_normal),
+                            derivation1.cmp(&deriv)
```
We have to be thankful to rust 1.29 for that. Otherwise we need to do an allocation by cloning derivation path.

sgeisler reviewed (changes requested):
I found some logic bugs which is probably a good indicator to add some tests for the newly added logic.

dr-orlovsky commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,52 @@ impl Map for Global {
+                        entry.insert((fingerprint1, derivation1));
+                        continue
+                    }
+                    else if derivation2[..] == derivation1[0..derivation2.len()]
```
I think I do not understand what we are trying to achieve here with the algorithm you proposed. So it's not just about writing more tests but about what we are doing with the merging at all

dr-orlovsky replied:
@sgeisler can you pls confirm that the new pushed version does what you meant by your algo?

sgeisler replied:
Thx for the quick changes :) I think we are now mostly talking about the same thing. But the slice arithmetic doesn't seem completely right yet. Will leave another comment.

sgeisler commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,52 @@ impl Map for Global {
+                        derivation1.len() < derivation2.len() &&
+                        derivation1[..] == derivation2[0..derivation1.len()]
+                    {
+                        entry.insert((fingerprint1, derivation1));
```
Yes, moving the insert into the next branch fixed it. I never meant we should insert derivation 2, you are right that we can just skip in that case.

sgeisler commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,54 @@ impl Map for Global {
+                    }
+                    else if
+                        len_diff_i64 < 0 &&
+                        derivation1[..] == derivation2[derivation2.len() - len_diff..]
```
Let's say `len1 = 2` and `len2 = 5`. That means `len_diff = 3`. This would make us compare `derivation1[0..2]` (of length 2) with `derivation2[2..5]` (of length 3). This doesn't seem to be  a valid suffix check as it would return false even though derivation 1 might be a suffix of derivation 2. 

I'd propose a more direct approach as I mentioned in another comment:

```rust
derivation1[..] == derivation2[(derivatio2.len()-derivation1.len())..]
```

I hope I didn't make any errors, I'm kinda tired but wanted to respond asap. So please double check.

dr-orlovsky replied:
Yes, this seems right. Sorry, was sleeping already. Pushed the latest version

sgeisler replied:
Thx, looks good now :)

sgeisler commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,54 @@ impl Map for Global {
+                    let (fingerprint2, derivation2) = entry.get().clone();
+
+                    let len_diff_i64 = derivation1.len() as i64 - derivation2.len() as i64;
+                    let len_diff = len_diff.abs() as usize;
```
I think you meant `let len_diff = len_diff_i64.abs() as usize;`?

dr-orlovsky replied:
Removed that part leaving explicit calculation in each case

dr-orlovsky commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,52 @@ impl Map for Global {
+                    //    - derivation paths are equal and fingerprints are not
+                    //    - derivation paths are of the same length, but not equal
+                    //    - derivation paths has different length, but the shorter one
+                    //      is not the strict prefix of the longer one
```
fixed

dr-orlovsky commented on src/util/psbt/map/global.rs:
```diff
@@ -113,6 +155,52 @@ impl Map for Global {
+                    }
+                    else if
+                        derivation1.len() < derivation2.len() &&
+                        derivation1[..] == derivation2[0..derivation1.len()]
```
fixed

apoelstra reviewed (approved):
ack 7f5c2795d6d581276119edf2becf4377d7c7df08
